# Overview

This repo contains the source code and structure to comprehensively describe
the analysis in [Thongdee et al 2025, *Modulation of protein activity by small
RNA base pairing internal to coding sequences* (Molecular Cell 85, 1-14)](https://doi.org/10.1016/j.molcell.2025.03.014).
Data are downloaded automatically from SRA, and config files are written to
work with these data. Other genomes or assemblies should be able to be used
with appropriate edits.

The complete set of
[Snakemake](https://snakemake.readthedocs.io/en/stable/) workflows,
configuration, supporting code, and downstream RMarkdown files are included.

## Requirements

- [conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html) must be installed
- this code repository must be downloaded and unpacked. Below, we refer to the
  unpacked directory as `$WORKDIR`. Unless otherwise specified, paths provided
  are assumed to be relative to `$WORKDIR`.

## Create conda environments

Create the conda environments in the top level of `$WORKDIR`:

```bash
conda env create -p env-rilseq --file env-rilseq.yml
```

## Run annotation workflow
The annotation workflow will create the annotation file
containing the CDS
features, 5UTR, 3UTR, AS (antisense) and IGR (intergenic regions) that are
needed for RILseq analysis. This workflow also downloads the genome reference fasta
file and generates the BWA-mem index necessary for mapping.
The `workflows/annotation/config/config.yaml`
file configures this.

To run the workflow, from the `workflows/annotation` directory, activate the
environment `env-rilseq` and run the Snakefile. Either use a [snakemake
profile](https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles)
appropriate for your cluster, or run on a single machine (specifying the number
of cores; here we use the placeholder `$CORES`)

```bash
# in the workflows/annotation directory
conda activate ../../env-rilseq  # activate environment
snakemake --cores $CORES  # run workflow
```

The main outputs are the RILseq-formatted annotation file,
`workflows/annotation/data/ecoli_default.patch1.gff`, and the bwa index file,
`workflows/annotation/data/ecoli_default.fasta.bwt`. These are already
configured to be used in other workflows. For example, for the
`workflows/rilseq/config/config.yaml` file, the entry `references:annotation`
points to the GFF file created by this workflow as
`../annotation/data/ecoli_default.patch1.gff`.

## Configure RIL-seq workflow

This can be found under `workflows/rilseq`. 

If necessary, edit the sample table (`workflows/rilseq/config/sampletable.tsv`).
This is currently configured to pull down data automatically from SRA, but if
running on different genomes or assemblies these would need additional
configuration.

Circos plots, generated by the rilseq workflow, need a configuration file that
matches the reference genome and annotation. Here we used the [default E. coli
file provided in the RILseq Github repo](https://github.com/asafpr/RILseq/blob/master/data/E_coli_K12/ver2/draw_links.conf). For
other organisms or assemblies, edit
`workflows/rilseq/config/RILseq-master/data/E_coli_K12/ver2/draw_links.conf` as
needed.

## Run RIL-seq workflow

This will first generate the cutadapt-trimmed FASTQ files using ``cutadapt``,
identify single and chimeric fragments and filter for significant chimeric
fragments using functions from the [RIL-seq
package](https://github.com/asafpr/RILseq). Plots and tables summarizing the
mapped fragments results are generated through downstream Rmarkdown scripts
which are also called from this workflow. Computation time takes several hours
per condition; samples are processed in batches per conditions. The RMarkdown
files in can take many hours to run for particuarly highly-sequenced samples.

Running this workflow, like the annotation workflow, needs the `env-rilseq`
environment activated.

```bash
# in workflows/rilseq:

conda activate ../../env-rilseq
snakemake --cores $CORES --use-conda
```

The main outputs of this workflow are the `all_fragments` and
`significant_fragments` files, stored in
`data/rnaseq_rilseq/{sample}/{sample}_all_fragments.txt` and
`data/rnaseq_rilseq/{sample}/{sample}_significant_fragments.txt` respectively.
These are TSVs of single and chimeric reads, the later interpreted as
interactions. They need to be further processed to map the fragments to
features, which is performed in the RMarkdown files.

This workflow also generates the Circos plots of interactions, stored in
`data/rnaseq_rilseq/{sample}/{sample}_circos_plot.svg`.

The output of the downstream analysis will be HTML files named after the
respective sample subgroup (i.e.,
`workflows/rilseq/data/downstream/log-noCL_summary.html` and
`workflows/rilseq/data/downstream/log-noCL_cds-shrink.html` for the `log-noCL`
subgroup). Documentation can be found with the HTML files. These RMarkdown
files also create Excel files with the summarized RIL-seq results.
